### PR TITLE
Withdrawals part 2

### PR DESCRIPTION
**Pull request by revitteth**
Continuation of PR #6009. Storage and retrieval of withdrawals.

- [x] storage of withdrawals on writeblock
- [x] composite key (block number | block hash)
- [x] value ([]withdrawal)
- [x] retrieval of withdrawals with block
- [x] tests around storing/retrieving withdrawals
- [x] tested in hive
- [x] commits tidied

## Hive Failures
- [x] Withdrawals Fork on Block 2
- [x] Withdrawals Fork on Block 3
- [ ] Sync after 2 blocks - Withdrawals on Block 1 - Single Withdrawal Account - No Transactions ( Syncing client rejected valid chain)
- [ ] Sync after 2 blocks - Withdrawals on Block 1 - Single Withdrawal Account (Timeout while waiting for secondary client to sync)
- [ ] Sync after 2 blocks - Withdrawals on Genesis - Single Withdrawal Account (Timeout while waiting for secondary client to sync)
- [ ] Sync after 2 blocks - Withdrawals on Block 2 - Single Withdrawal Account - No Transactions (Syncing client rejected valid chain)
- [ ] Sync after 2 blocks - Withdrawals on Block 2 - Single Withdrawal Account (Timeout while waiting for secondary client to sync)
- [ ] Sync after 128 blocks - Withdrawals on Block 2 - Multiple Withdrawal Accounts (Unexpected error on BalanceAt: Post "http://172.17.0.4:8545/": context deadline exceeded, expected=<None>)
- [ ] Withdrawals Fork on Block 1 - 8 Block Re-Org, Sync (Expected no error on EngineNewPayloadV2: error=Post "http://172.17.0.4:8551/": context deadline exceeded)
- [ ] Withdrawals Fork on Block 8 - 10 Block Re-Org Sync (Expected no error on EngineNewPayloadV2: error=Post "http://172.17.0.4:8551/": context deadline exceeded)
- [ ] Withdrawals Fork on Canonical Block 8 / Side Block 7 - 10 Block Re-Org Sync (Expected no error on EngineNewPayloadV2: error=Post "http://172.17.0.4:8551/": context deadline exceeded)
- [ ] Withdrawals Fork on Canonical Block 8 / Side Block 9 - 10 Block Re-Org Sync (Expected no error on EngineNewPayloadV2: error=Post "http://172.17.0.4:8551/": context deadline exceeded)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -72,6 +72,7 @@ tests/testdata/*
 
 go.work
 go.work.sum
+erigon-lib/
```
Comment (revitteth) [inline .gitignore:75]: This lets us do docker stuff (keeps erigon-lib within permissible context).

File: core/rawdb/accessors_chain.go [hunk 2]
```diff
@@ -686,9 +689,13 @@ func WriteBody(db kv.RwTx, hash common.Hash, number uint64, body *types.Body) er
 		TxAmount: uint32(len(body.Transactions)) + 2,
 		Uncles:   body.Uncles,
 	}
+	if len(body.Withdrawals) > 0 {
```
Comment (yperbasis) [inline core/rawdb/accessors_chain.go:692]: It should be `if body.Withdrawals != nil {` instead in order to preserve the difference between nil and empty withdrawals.

**Review by yperbasis — CHANGES_REQUESTED** [review]

File: core/types/block.go [hunk 3]
```diff
@@ -535,10 +535,10 @@ type RawBody struct {
...
 
 type BodyForStorage struct {
-	BaseTxId uint64
-	TxAmount uint32
-	Uncles   []*Header
-	// TODO(yperbasis): withdrawals
+	BaseTxId    uint64
+	TxAmount    uint32
+	Uncles      []*Header
+	Withdrawals *[]*Withdrawal `rlp:"optional,nil"`
```
Comment (yperbasis) [inline core/types/block.go:541]: Will the rlp optional logic work if the type is changed to `[]*Withdrawal` instead of `*[]*Withdrawal`?

Comment (revitteth) [reply core/types/block.go:541]: Unfortunately we lose the functionality of nil if we don't use a pointer type

File: core/rawdb/accessors_chain.go (hunk 2 above)
Comment (revitteth) [reply core/rawdb/accessors_chain.go:692]: Good point!

File: core/rawdb/accessors_chain_test.go [hunk 4]
```diff
@@ -434,6 +435,141 @@ func TestBlockReceiptStorage(t *testing.T) {
...
+	if err := WriteBlock(tx, block); err != nil {
+		t.Fatalf("Could not write block: %v", err)
+	}
+	if _, _, err := DeleteAncientBlocks(tx, 0, 1); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// Tests pre-shanghai body to make sure withdrawals doesn't panic
+func TestPreShanghaiBodyNoPanicOnWithdrawals(t *testing.T) {
```
Comment (yperbasis) [inline core/rawdb/accessors_chain_test.go:560]: Please add a similar test with pre-existing RLP for BodyForStorage.

File: core/types/block.go (hunk 3 above)
Comment (yperbasis) [reply core/types/block.go:541]: For normal slices in go there's a difference between a nil slice and and empty slice – you don't need a pointer for that.

File: core/rawdb/accessors_chain.go [hunk 5]
```diff
@@ -488,11 +489,13 @@ func WriteRawTransactions(tx kv.RwTx, txs [][]byte, baseTxId uint64) error {
 
 // WriteBodyForStorage stores an RLP encoded block body into the database.
 func WriteBodyForStorage(db kv.Putter, hash common.Hash, number uint64, body *types.BodyForStorage) error {
-	data, err := rlp.EncodeToBytes(body)
```
Comment (yperbasis) [inline core/rawdb/accessors_chain.go:491]: Why change the old code.

File: core/rawdb/accessors_chain.go [hunk 6]
```diff
@@ -604,7 +607,8 @@ func ReadBodyForStorageByKey(db kv.Getter, k []byte) (*types.BodyForStorage, err
 		return nil, nil
 	}
 	bodyForStorage := new(types.BodyForStorage)
-	if err := rlp.DecodeBytes(bodyRlp, bodyForStorage); err != nil {
```
Comment (yperbasis) [inline core/rawdb/accessors_chain.go:607]: Again, why change the old code?

File: core/rawdb/accessors_chain.go [hunk 7]
```diff
@@ -617,13 +621,17 @@ func ReadBody(db kv.Getter, hash common.Hash, number uint64) (*types.Body, uint6
...
-	err := rlp.DecodeBytes(data, bodyForStorage)
+	s := rlp.NewStream(bytes.NewReader(data), 0)
+	err := bodyForStorage.DecodeRLP(s)
 	if err != nil {
 		log.Error("Invalid block body RLP", "hash", hash, "err", err)
 		return nil, 0, 0
 	}
 	body := new(types.Body)
 	body.Uncles = bodyForStorage.Uncles
+	if bodyForStorage.Withdrawals != nil {
```
Comment (yperbasis) [inline core/rawdb/accessors_chain.go:632]: Unnecessary `!= nil `check

File: core/rawdb/accessors_chain.go [hunk 8]
```diff
@@ -682,13 +690,18 @@ func WriteBody(db kv.RwTx, hash common.Hash, number uint64, body *types.Body) er
...
 	data := types.BodyForStorage{
-		BaseTxId: baseTxId,
-		TxAmount: uint32(len(body.Transactions)) + 2,
-		Uncles:   body.Uncles,
+		BaseTxId:    baseTxId,
+		TxAmount:    uint32(len(body.Transactions)) + 2,
+		Uncles:      body.Uncles,
+		Withdrawals: nil,
+	}
+	if body.Withdrawals != nil {
```
Comment (yperbasis) [inline core/rawdb/accessors_chain.go:698]: Unnecessary `!= nil `check

File: core/types/block.go [hunk 9]
```diff
@@ -746,6 +746,151 @@ func (rb *RawBody) DecodeRLP(s *rlp.Stream) error {
...
+	}
+	// end of Uncles
+	if err = s.ListEnd(); err != nil {
+		return err
+	}
+
+	// decode Withdrawals
+	if _, err = s.List(); err != nil {
+		if errors.Is(err, rlp.EOL) {
+			bfs.Withdrawals = []*Withdrawal{}
```
Comment (yperbasis) [inline core/types/block.go:867]: Should be `bfs.Withdrawals = nil` instead

File: core/types/block.go [hunk 10]
```diff
@@ -1098,7 +1243,7 @@ func (bb *Block) DecodeRLP(s *rlp.Stream) error {
 		return err
 	}
 
-	return s.ListEnd()
+	return nil
```
Comment (yperbasis) [inline core/types/block.go:1246]: This change looks like a mistake. We should end not only the list of withdrawals, but the main body list as well.

File: core/types/block.go [hunk 11]
```diff
@@ -746,6 +746,151 @@ func (rb *RawBody) DecodeRLP(s *rlp.Stream) error {
...
+	}
+	if !errors.Is(err, rlp.EOL) {
+		return err
+	}
+	// end of Withdrawals
+	if err = s.ListEnd(); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (yperbasis) [inline core/types/block.go:891]: Should be return `s.ListEnd()` instead.

File: core/types/block.go [hunk 12]
```diff
@@ -746,6 +746,151 @@ func (rb *RawBody) DecodeRLP(s *rlp.Stream) error {
...
+		return err
+	}
+
+	// decode Withdrawals
+	if _, err = s.List(); err != nil {
+		if errors.Is(err, rlp.EOL) {
+			bfs.Withdrawals = []*Withdrawal{}
+			return s.ListEnd()
+		}
+		if errors.Is(err, io.EOF) {
```
Comment (yperbasis) [inline core/types/block.go:870]: Don't think it's necessary to handle `io.EOF`.

**Review by yperbasis — CHANGES_REQUESTED** [review]

File: core/types/block.go (hunk 10 above)
Comment (revitteth) [reply core/types/block.go:1246]: Maybe this is an issue with how I'm encoding as RLP, I've encoded the 2 ints and then the 2 lists, there isn't a container list for all 4 properties on the struct.

Comment (yperbasis) [reply core/types/block.go:1246]: Yeah, you need to call `EncodeStructSizePrefix(payloadSize, ...)` first. Use `(rb RawBody) EncodeRLP(w io.Writer)` as an example.

**Comment by revitteth** [comment]
@yperbasis there are still some failing hive tests, but this is close to ready I think! I will investigate the remaining failures.

File: core/types/block.go (hunk 10 above)
Comment (yperbasis) [reply core/types/block.go:1257]: The last line of `func (bb *Block) DecodeRLP` should stay `return s.ListEnd()`

File: eth/protocols/eth/protocol.go [hunk 13]
```diff
@@ -555,6 +566,24 @@ func (bb *BlockBody) DecodeRLP(s *rlp.Stream) error {
 	if err = s.ListEnd(); err != nil {
 		return err
 	}
+	// decode Withdrawals
+	if _, err = s.List(); err != nil {
```
Comment (yperbasis) [inline eth/protocols/eth/protocol.go:570]: Please add something like
```
		if errors.Is(err, rlp.EOL) {
			bb.Withdrawals = nil
			return s.ListEnd()
		}
```
in order to handle pre-Shanghai block bodies.

File: eth/protocols/eth/protocol.go [hunk 14]
```diff
@@ -555,6 +566,24 @@ func (bb *BlockBody) DecodeRLP(s *rlp.Stream) error {
...
+		return err
+	}
```
Comment (yperbasis) [inline eth/protocols/eth/protocol.go:572]: Please add
```
bb.Withdrawals = []*Withdrawal{}
```
in order to handle post-Shanghai bodies with empty withdrawals (see PR #6291).

**Review by yperbasis — CHANGES_REQUESTED** [review]

**Review by yperbasis — APPROVED** [review]

**Pull request merged**